fastly/resource_fastly_ngwaf_virtualpatches.go

```diff
@@ -0,0 +1,157 @@
+package fastly
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+
+	gofastly "github.com/fastly/go-fastly/v11/fastly"
+	ws "github.com/fastly/go-fastly/v11/fastly/ngwaf/v1/workspaces/virtualpatches"
+)
+
+func resourceFastlyNGWAFVirtualPatches() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceFastlyNGWAFVirtualPatchUpdate,
+		DeleteContext: resourceFastlyNGWAFVirtualPatchDelete,
+		ReadContext:   resourceFastlyNGWAFVirtualPatchRead,
+		UpdateContext: resourceFastlyNGWAFVirtualPatchUpdate,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceFastlyNGWAFVirtualPatchImport,
+		},
+		Schema: map[string]*schema.Schema{
+			"action": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "User-configured action of the virtual patch. Accepted values are [`block`, `log`]",
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice(
+					[]string{"log", "block"},
+					false,
+				)),
+			},
+			"enabled": {
+				Type:        schema.TypeBool,
+				Required:    true,
+				Description: "User-configured control for enabling or disabling virtual patches",
+			},
+			"virtual_patch_id": {
+				Type:             schema.TypeString,
+				Required:         true,
+				Description:      "User-submitted identifier (ID) of the virtual patch",
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotWhiteSpace),
+			},
+			"workspace_id": {
+				Type:             schema.TypeString,
+				Required:         true,
+				Description:      "User-submitted identifier (ID) of the workspace",
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotWhiteSpace),
+			},
+		},
+	}
+}
+
+func resourceFastlyNGWAFVirtualPatchRead(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	conn := meta.(*APIClient).conn
+
+	i := ws.GetInput{
+		WorkspaceID:    gofastly.ToPointer(d.Get("workspace_id").(string)),
+		VirtualPatchID: gofastly.ToPointer(d.Get("virtual_patch_id").(string)),
+	}
+
+	log.Printf("[DEBUG] REFRESH: NGWAF virtual patch input: %#v", i)
+
+	virtualpatch, err := ws.Get(gofastly.NewContextForResourceID(ctx, d.Id()), conn, &i)
```

Comment:
The ID passed to `NewContextForResourceID` must be the workspace ID, not the ID of the Virtual Patch resource.